internal/fxreflect/fxreflect.go

```diff
@@ -25,25 +25,84 @@ import (
 	"reflect"
 	"runtime"
 	"strings"
+
+	"go.uber.org/dig"
 )
 
-// ReturnTypes takes a func and returns a slice of string'd types
-// TODO instead of duplicating the dig's reflect logic, trying to
-// determine which types actually made it into the container, have
-// dig return a Result struct which could contain the ProvidedTypes
+// ReturnTypes takes a func and returns a slice of string'd types.
 func ReturnTypes(t interface{}) []string {
+	// TODO(grayson): instead of duplicating the dig's reflect logic, trying to
+	// determine which types actually made it into the container, have
+	// dig return a Result struct which could contain the ProvidedTypes
+
+	if reflect.TypeOf(t).Kind() != reflect.Func {
+		// Invalid provide, will be logged as an error.
+		return []string{}
+	}
+
 	rtypes := []string{}
-	fn := reflect.ValueOf(t).Type()
+	ft := reflect.ValueOf(t).Type()
 
-	for i := 0; i < fn.NumOut(); i++ {
-		if !isErr(fn.Out(i)) {
-			rtypes = append(rtypes, fn.Out(i).String())
-		}
+	for i := 0; i < ft.NumOut(); i++ {
+		t := ft.Out(i)
+
+		traverseOuts(key{t: t}, func(s string) {
+			rtypes = append(rtypes, s)
+		})
 	}
 
 	return rtypes
 }
 
+// this type is basically straight out of dig, which is a strong signal
+// that exporting it could really DRY up some things for fx-dig relationship.
```

Comment:
I think that bringing logging into dig would do *more* rather than *less*. 

Dig is not meant for direct consumption, at least that's how we've been thinking about it and operating it. Fx is the user-facing product and should care about things like exposing a sane logger, or the ability to override it.

What I think we all agree on is that dig should allow for more introspection, without necessarily tying it to a particular use-case (logging, instrumentation, etc.)

If we want to try something something really simple, we could start by exposing an "action" channel as a `New` option, which will send notifications when a type is added, or invoked.

(Lets not focus on the names too much right now, rather than the concept)

```go
const (
  Provided = iota
  Invoked
)

type Key struct { // export existing structure
  t reflect.Type
  name string

type Action struct {
  Key

  ActionType int
}
```

Consumption side would look something like:

```go
actions := make(chan dig.Action)
c := dig.New(dig.Listener(actions))
// spin up goroutine to log actions
```

Happy to move this out of the PR into an issue discussion on dig, or to sync up today.